File: frontend/src/protocol.js
export const MAX_PACKET_BYTES = 262144;
const object = value => value !== null && typeof value === 'object' && !Array.isArray(value);
function jsonValue(value, depth = 0) {
  if (depth > 32) throw Error('JSON nesting too deep');
  if (value === null || typeof value === 'string' || typeof value === 'boolean') return;
  if (typeof value === 'number' && Number.isFinite(value)) return;
  if (Array.isArray(value) || object(value)) {
    for (const child of Object.values(value)) jsonValue(child, depth + 1);
    return;
  }
  throw Error('Invalid JSON value');
}
export function validatePacket(input) {
  if (typeof input === 'string' && new TextEncoder().encode(input).length > MAX_PACKET_BYTES) throw Error('Packet too large');
  const p = typeof input === 'string' ? JSON.parse(input) : input;
  if (!object(p) || p.version !== 1) throw Error('Unsupported envelope');
  for (const key of ['type', 'source', 'session_id']) {
    if (typeof p[key] !== 'string' || [...p[key]].length < 1 || [...p[key]].length > 128) throw Error(`Invalid ${key}`);
  }
  if (!Number.isSafeInteger(p.sequence) || p.sequence < 0) throw Error('Invalid sequence');
  if (typeof p.timestamp !== 'number' || !Number.isFinite(p.timestamp) || p.timestamp < 0 || p.timestamp > Number.MAX_SAFE_INTEGER) throw Error('Invalid timestamp');
  if (!object(p.payload)) throw Error('Invalid payload');
  jsonValue(p);
  const encoded = JSON.stringify(p);
  if (new TextEncoder().encode(encoded).length > MAX_PACKET_BYTES) throw Error('Packet too large');
  return JSON.parse(encoded);
}
export function validateSnapshot(value) {
  if (!object(value) || !Number.isSafeInteger(value.sequence) || value.sequence < 0 ||
      !(value.session_id === null || (typeof value.session_id === 'string' && value.session_id.length > 0)) ||
      !Array.isArray(value.packets) || value.packets.length > 128) throw Error('Invalid snapshot');
  const packets = value.packets.map(validatePacket);
  let last = -1;
  const keys = new Set();
  for (const p of packets) {
    const key = JSON.stringify([p.source, p.type]);
    if (p.session_id !== value.session_id || p.sequence <= last || p.sequence > value.sequence || keys.has(key)) throw Error('Inconsistent snapshot');
    last = p.sequence; keys.add(key);
  }
  return { sequence: value.sequence, session_id: value.session_id, packets };
}

File: backend/app/protocol.py
"""Versioned, payload-agnostic transport. Times use session-relative seconds."""
import json
import math

MAX_PACKET_BYTES = 262144


def _json_value(value, depth=0):
    if depth > 32:
        raise ValueError('JSON nesting too deep')
    if value is None or type(value) in (bool, str, int):
        return
    if type(value) is float and math.isfinite(value):
        return
    if isinstance(value, list):
        for child in value:
            _json_value(child, depth + 1)
        return
    if isinstance(value, dict) and all(isinstance(key, str) for key in value):
        for child in value.values():
            _json_value(child, depth + 1)
        return
    raise ValueError('packet must contain finite JSON values with string keys')


def validate_packet(packet):
    if not isinstance(packet, dict):
        raise ValueError('packet must be an object')
    if type(packet.get('version')) is not int or packet['version'] != 1:
        raise ValueError('unsupported version')
    for key in ('type', 'source', 'session_id'):
        if not isinstance(packet.get(key), str) or not 0 < len(packet[key]) <= 128:
            raise ValueError('invalid ' + key)
    if type(packet.get('sequence')) is not int or not 0 <= packet['sequence'] <= 9007199254740991:
        raise ValueError('invalid sequence')
    timestamp = packet.get('timestamp')
    if type(timestamp) not in (int, float) or not 0 <= timestamp <= 9007199254740991:
        raise ValueError('invalid timestamp')
    if not isinstance(packet.get('payload'), dict):
        raise ValueError('payload must be an object')
    _json_value(packet)
    try:
        encoded = json.dumps(packet, allow_nan=False)
    except (ValueError, TypeError) as exc:
        raise ValueError('packet must contain finite JSON values') from exc
    if len(encoded.encode()) > MAX_PACKET_BYTES:
        raise ValueError('packet too large')
    return json.loads(encoded)  # Detach mutable producer objects.


def make_packet(kind, timestamp, sequence, source, session_id, payload):
    return validate_packet(dict(version=1, type=kind, timestamp=timestamp,
                                sequence=sequence, source=source, session_id=session_id, payload=payload))
